actions: compliance: Add support for GitHub annotations

carlescufi opened the pull request:
Add support for annotations as described:
https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#about-workflow-commands

Note: DNM because I will be using this PR to test it.

carlescufi committed:
scripts: compliance: Rename add_failure to failure

For consistency with error() and skip(), rename add_failure() to
failure().

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
scripts: compliance: Add a new formatted failure method and class

Format the JUnit failure according to the standard:
https://www.ibm.com/docs/en/developer-for-zos/14.2?topic=formats-junit-xml-format

This also includes a new FmtdFailure class, which will be used in
subsequent commits to implement GitHub-compatible annotations. A list of
FmtdFailure instances is kept in the ComplianceTest instance because
the TestCase instance cannot restore classes derived from Result that
the junitparser module doesn't know about.

Use the new function to format checkpatch errors.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
scripts: compliance: Add GitHub-compatible annotation support

Add support for GitHub workflow annotations, so that errors are
displayed inline with the Pull Request diff.

More info here:
https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#about-workflow-commands

Enable the new option in the compliance GitHub action.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

gmarull commented on scripts/ci/check_compliance.py:
```diff
@@ -990,6 +1033,18 @@ def inheritors(klass):
+    """
+    https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#about-workflow-commands
+    """
+    if not isinstance(res, FmtdFailure):
+        return
```
is this check needed?

carlescufi replied:
nope, will remove, it's a leftover, thanks.

carlescufi committed:
scripts: compliance: Modernize pylint invocation

The current recommended API call for subprocess instantiation is run(),
so replace the low-level Popen and communicate() with it.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
scripts: compliance: Parse the pylint output

Parse the output generated by pylint and store it as FmtdFailure
instances, so that they can be annotated later.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

galak commented:
What does annotations get us?

nashif commented:
> What does annotations get us?

this: 
![image](https://user-images.githubusercontent.com/180017/201180588-0e650679-7692-4a76-a943-8d68ccbce270.png)


You do not have to look at the action log and figure out what the text output is talking about, the comments are directly placed where the violation happens.

carlescufi committed:
requirements: Add lxml as a requirement

junitparser has a bug that prevents it from working correctly when lxml
is not installed.

See this GH issue for more information:
https://github.com/weiwei/junitparser/issues/99

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
scripts: compliance: Ensure lxml is installed

See https://github.com/weiwei/junitparser/issues/99.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>
